### ISSUE TITLE
Video streaming uses much higher CPU for streaming in v10.0.0-beta1

### ISSUE THREAD

**Issue by tsightler** [dgreif/ring#902]
# Bug Report

### Describe the Bug
Hi @dgreif,

I wasn't sure if I should just comment in the other thread, but I thought it would be worth opening a new thread instead.  I've had pretty good success with the v10 beta of ring-client-api, but one thing that I have noticed is that the node process uses significantly higher CPU than the older streaming method.

For example, with 9.24.0, I can support about 7 actiive streams on an RPi4b with only some minor artifacts, 6 streams works pretty flawlessly.  CPU usage on the Node process approaches 100% but it handles it OK.

However, with v10.0.0, I can support about half as many streams on the same system.  3 streams seems to work OK, but if I start a 4th stream, CPU usage goes to 100% and artifacts begin to appear in the streams due to lack of CPU.  I'm curious if you have any ideas on why this might be.  The ffmpeg process are configured to just copy the stream to an RTSP server, no transcoding or anything, so the use for those per-stream is still quite low, it seems to be higher CPU usage in the RTP packet handling within the Node process itself.

I may be able to work around this by forking the streams off to separate node processes as it seems that the other cores have some CPU headroom, and perhaps I should do this anyway, but as it was demonstratably different with just chaning the API version back and forth I thought it might be worth mentioning as it was a surprise to me with the new streaming method that otherwise works exceptionally well.  It looks like the new code completely replaces the RTP packet handling code, so perhaps something in there isn't as efficient as the old code?

Ideally I'd love to figure out how to just consume the RTP streams directly, bypassing node for that part, and I feel like this should be possible, but I haven't had time to dig into it that much.  Happy to provide any more detail if it helps.

### To Reproduce
Just call camera.streamVideo() for multiple streams and monitor CPU usage with 9.24.0 vs 10.0.0-beta1

### Expected behavior
Similar CPU usage to prior API versions

### Environment
 - OS: Alpine container on HASS OS and Ubuntu 20.04
 - Node.js: 16.13.2

**Comment by tsightler** [comment]
So, I implemented a crude, but functional, multi-threaded solution by modifying my code to create a worker thread for each camera during startup.  This definitely made a difference, obviously not by lowering the CPU usage per stream, but just by increasing the ability of the code to scale to more cores.  This made it mostly work with 5 streams, but requires 75-80% of the entire CPU of the RPi4 vs using less than 50% previously with 7 streams.  I could sometimes get a 6th stream to work but with a lot of artifacting due to high CPU usage.

I'm not sure if there's much that can be done about it in ring-client-api itself as it looks like the RTP packet processing is mostly in werift.   Since I don't really need return audio I'm thinking the ideal case would be for me to figure out how to just use the SDP info to have ffmpeg connect to the stream directly, but I don't know if I have enough time to work through that approach, or if it's even workable.

**Comment by dgreif** [comment]
I wonder if part of it is the video profile that we are receiving from Ring.  I think it's using the High profile now by default. I can't currently request anything lower with v10, but I've got some changes in the works that would allow us to request whatever video profile we would like.  Will keep you posted.

**Comment by tsightler** [comment]
Thanks @dgreif, I had wondered about that as well as it seemed like there was more bandwidth usage which likely translates to more CPU, but I haven't really had time to dig at it in more detail and be sure.  I'll anxiously await your improvements.

In the interim, it's still made me realize that I should use worker threads for each stream to maximize performance, especially on these lower end devices, so I'm working on a better implementation of that.

**Comment by tsightler** [comment]
I did some very light profiling between these two versions and, at first glance, it seems the difference is perhaps in decryption.  With the WebRTC based API the NodeJS decrypt calls are clearly one of the highest CPU consumers in the pipeline, but when looking at the profile for a SIP based stream, I don't even see decrypt in the profile.  The overall packet numbers look basically identical between the two, so I'm less convinced it's the profile, although the profile might make a difference.

I admittedly need to setup a better test as this was not a 100% 1:1 comparison as I was comparing my dev branch to main, but the overall code path for streams is the same and I thought I'd share this initial feedback in case it triggers any other thoughts.  Does WebRTC use encryption of the RTP streams by default while SIP does not perhaps?  I'm going to setup a more 1:1 test scenario and dig a little deeper.

**Comment by tsightler** [comment]
Attaching some CPU profiles from identical code streaming from that same camera with minimal motion for both SIP and WebRTC based API.  Profile was captured for ~60 seconds (I think 57 and 55 seconds actually, my timing was a little off).  You can see the clear increase in time even though these were captured running on a modern AMD Ryzen processor, not an ARM , which is where the problem is significantly more pronounced.

The only significant difference seems to be that about 80% of the increase in time (maybe even more with overhead) is in the calls to decrypt of the srtp streams.  Perhaps this is the big difference and maybe nothing can be done about it but it's a pretty big issue as it makes using the code on a RPi less than ideal.

I'm not sure about profiles, to me this looks like baseline profile, I wonder if high might actually help since it should reduce bandwidth and I'm not actually processing the stream, just copying it.  I need to play with it more, not really drawing any conclusions yet, just thought the comparison profiles might be marginally interesting.

[sip-stream.zip](https://github.com/dgreif/ring/files/8247894/sip-stream.zip)
[webrtc-stream.zip](https://github.com/dgreif/ring/files/8247895/webrtc-stream.zip)

**Comment by tsightler** [comment]
I've been digging through the werift code trying to understand it better and I think I've come to the conclusion that decrypting SRTP in native Javascript is where the extra CPU is coming from but, more critically, it adds just enough latency to lower the total packets per second processing of the pipeline below the threshold when bitrates approach the high end of the variable bitrate.  If it's truly possible to lower the selected profile, either limiting the bitrate or even negotiating a lower resolution (I think I've seen 720p sometimes), then that might make a difference, but more by lowering total number of RTP packets to process than anything else.

Since Scrypted also uses werift and has a similar on-demand RTSP capability as ring-mqtt, I decided to play around with it and found that the CPU usage and artifacting follows a nearly identical pattern when used on the same hardware.  This at least helps me know there's not much I can do about it in my existing pipeline.

I had the idea to modify werift to bypass the decryption and send the SRTP packets directly to FFMPEG.  I figured out how to extract the session keys/salt from the SRTP session context after they are negotiated by DTLS and I modified werift to skip the decryption phase for SRTP so that I could send the packets unmodified to ffmpeg, but I failed in my efforts to get this to actually work.  I may try again when I have more time or maybe I should just finally break down and write a small go tool using Pion-WebRTC which would probably perform quite a bit better on these lower end CPUs.

**Comment by dgreif** [comment]
@tsightler I just released `10.0.0-beta.2`, which requests the same h264 profile as v9.  It's still not extremely reliable when connecting to certain cameras, so keep apologies if not all of your streams work with this update.  Let me know how it looks from a CPU perspective

**Comment by tsightler** [comment]
Thanks @dgreif, I'll give it a spin.  I had already hacked around with a custom version of the APi to play with profiles and hadn't had much success.  I did wonder if I could request some lower bitrate though.

**Comment by tsightler** [comment]
Hi @dgreif.  I made the required modifications to my worker thread implementation so that the beta2 code would work (for previous code I just generated sessionId from main thread and passed, but new code requires access_token and sessionId is generated within Live-call process on API so it required some slight reworking to the IPC code, but nothing too deep).

Unfortunately, as I suspected, it didn't help the CPU usage at all and has actually introduced a new problem as the new code seems to take quite a bit longer, in relative terms, before the channels actually start sending data.  I'm seeing an average of around 1 second of extra delay (3-4 seconds instead of 2-3 seconds) before the streams really begin rolling.   This doesn't sound like much, but it seems to be just enough that the frontend on Home Assistant times out the RTSP stream just before the stream starts, while before, this would only happen only in very rare cases.

Just to be clear, the CPU usage itself isn't too much of the issue, but the problem is that, sometimes even with just one stream, it's dropping data enough to corrupt the image.  I've included a screenshot of one of my front floodlight cameras to give an idea.  I'm thinking that perhaps this could be overcome by increasing the UDP buffer, but, unfortunately, the use case for most users requires running this code in a container on Home Assistant OS and the container environment doesn't have access to increase this and accessing the base OS is difficult for most users.  Note that it works perfectly on both my dev workstation and in a VM on ESXi, so it's just the RPi where there are problems.  It seems like it takes just enough time that, for a complex image, the latter SRTP packets are dropped.  If the image is more consistent (for example at night with night vision), so there's less RTP packets overall, the code works quite will.

I've included a couple of screenshots showing the CPU usage with the same two cameras streaming with 9.x vs 10.x just to give an idea of what I'm talking about.  It's about 2x the CPU, and, overall, it doesn't look that bad at first glance, as it only moves from 10% CPU to 20% CPU with two streams, but of course that 20% of 4 cores in the RPi4b and, the more threads there are, the more overhead.

It seems that most of the "extra" CPU usage is in the SRTP decryption, which has to calculate a new IV for each packet, then call createDecpheriv() for each one.  Again, it doesn't sound like much, but, as the loop was already right on the edge, these extra few ms per packet are really noticable.  Maybe there's something that could be improved in werift, a bigger inbound buffer or something, not sure.

For now I'm investigating other options as I'm guessing that WebRTC is going to be the way forward no matter what, especially looking at all the Ring Edge changes in there.  There is some WebRTC compatibility going into Home Assistant, so maybe at some point I can just pass off the WebRTC info there vs having to pass the streams through ffmpeg/RTPS.  This seems to be what Scrypted does when users are viewing the streams via the web browser, which makes total sense, no need to proxy that data through the RPi.

IN the meantime I'm looking at possibly wrapping Gstreamer and its webrtcbin, which would let me just do the required audio transcoding and even the RTSP server straight from a single pipeline, or perhaps reimplementing the basic livecall logic from ring-client-api into python using aiortc instead (it hands off the entire SRTP path to C code, not just the decipher) or perhaps writing my own handler in Go/Rust using Pion-WebRTC.

<img width="820" alt="ring-stream-api-9x" src="https://user-images.githubusercontent.com/17784731/160486417-504f441b-01b4-4ba7-90ff-10c0374a2d03.png">
<img width="823" alt="ring-stream-ap
<img width="770" alt="corrupt-image
<img width="769" alt="good-image" src="https://user-images.githubusercontent.com/17784731/160486442-0f82d305-f8dd-4ddd-b7c0-e3bd00554a19.png">
" src="https://user-images.githubusercontent.com/17784731/160486435-8d5984d2-7e91-44e1-ae69-8252761b43c8.png">
i-10x" src="https://user-images.githubusercontent.com/17784731/160486422-2cff4106-b5e3-407c-91d5-0ae5b518780f.png">

**Comment by dgreif** [comment]
@tsightler thanks for continuing to dig into this issue!  I'll try to get the beta wrapped up soon, and one thing I plan to add is an option to override the `PeerConnection` class so that you can provide a custom webrtc solution.  Scrypted does allow streaming to the browser, and that's one use case where this would be really useful.

**Comment by tsightler** [comment]
I have verified that increasing the default rmem_max/rmem_default largely makes the video artifacting shown in the image posted above go away and can clearly see that the core problem is that werift is unable to process the incoming bursts of RTP traffic before they overrun the default Linux buffer which is a paltry 200K, not much for modern day streaming.  When a stream first starts you'll almost always see 400-500K of receive queue buffers before the pipeline really gets rolling.  Without increasing the buffers these early packets were always lost, causing delay for stream startup.

Even a small increase in buffer size to 1M makes a huge difference in the reliability of the stream on the RPi4 and I get pretty much zero artifacts with 4 streams.  You will see the receive queues build up occasionally, but werift will eventually catches up and is able to largely maintain the stream, but there are regular bursts where the receive queue is well above 200K, which is why the smaller buffer size caused artifacts even sometimes with one camera.

At 5 streams werift begins to really struggle and 1M isn't enough and you start to see some of the receive queues overrun.  If I increase the rmem_max/rmem_default to something larger like 25M, then even 5 or 6 streams will mostly work, and I even managed to start a 7th stream, but werift can't process the packets as fast as they are arriving and queues slowly get larger and larger.  I don't consider that really a problem though as I couldn't support 7 concurrent streams with the old SIP code either, 5-6 was about the limit.

So, basically, once I manually increase the Linux buffers, the beta can support roughly the same number of streams as the prior code, just with 2x more overall CPU.  For users that run this on anything more powerful than an RPi it's probably just going to be fast enough anyway (I was able to easily have 7 active streams on a 2-vCPU VM running an Intel processor from 9 years ago), and those that either manually install this, or use it in Docker but manage the host OS themselves, increasing the buffers is a pretty easy task so I can just include some instructions for those users.

Unfortunately for me, most users of ring-mqtt these days consume the project as an addon for Home Assistant, which intentionally makes accessing the actual host OS difficult.  It's possible to do so by creating an authorized_keys file and moving it over via USB, but that's really intended for developers and I'm not sure how persistent changes made there really are, but I'll investigate.  I think I'll open an issue on the HassOS project and see if they will consider simply increasing the default values slightly since I see they already have a few tweaks for multicast and streaming is a much bigger part of Home Assistant these days so it won't hurt anything.

In the end, none of this really has anything to do with this project directly, so I'm going to close this as the code works as designed.  @dgreif I really appreciate your efforts to keep this API updated and supporting all the latest Ring APIs while keeping the changes to those of us that use the API to a minimum.  The ring-mqtt project wouldn't be possible without this as I barely have time to keep it going as it is!  Thanks again!
